### PR TITLE
GPUvsCPU: updates to modules and workflows for pixel reconstruction

### PR DESCRIPTION
**Pull request by AdrianoDee**
#### PR description:

Updates for pixel only workflows:
- rearranging `.502` and `.503` to avoid running the CPU chain when a GPU is available (due to the fact that the DQM consumes hit SoA and that is only available on CPU);
- introducing a module to copy GPU SoA to CPU;
- ~running both CPU and GPU chains in `.503`;~
- ~introducing `gpuPixelValidation` modifier;~
- ~dropping `.501` that only generates confusion.~

#### PR validation:

Running `.502` and `.503`.

### CODE REVIEW

**Comment by AdrianoDee** [comment]
enable gpu

File: CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h [hunk 1]
```diff
@@ -178,4 +204,74 @@ TrackingRecHit2DHeterogeneous<Traits>::TrackingRecHit2DHeterogeneous(
...
+  m_store16 = Traits::template make_unique<uint16_t[]>(nHits * n16, stream);
+  m_store32 = Traits::template make_unique<float[]>(nHits * n32, stream);
+  m_PhiBinnerStore = Traits::template make_unique<TrackingRecHit2DSOAView::PhiBinner>(stream);
+  m_AverageGeometryStore = Traits::template make_unique<TrackingRecHit2DSOAView::AverageGeometry>(stream);
+
+  view->m_averageGeometry = m_AverageGeometryStore.get();
+  view->m_hitsModuleStart = m_hitsModuleStart;
+
+  //store transfer
+  if constexpr (std::is_same<Traits, cms::cudacompat::GPUTraits>::value) {
```
Comment (fwyzard) [inline CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h:236]: just a suggestion:
```suggestion
  if constexpr (std::is_same_v<Traits, cms::cudacompat::GPUTraits>) {
```

**Comment by fwyzard** [comment]
The c++ part looks good to me :-)
Now to understand the python part ...

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

File: DQM/SiPixelPhase1Heterogeneous/python/SiPixelPhase1HeterogenousDQM_FirstStep_cff.py [hunk 2]
```diff
@@ -1,10 +1,30 @@
...
+siPixelPhase1MonitorVertexSoA = _siPixelPhase1MonitorVertexSoA.clone()
+siPixelPhase1MonitorTrackSoA = _siPixelPhase1MonitorTrackSoA.clone()
 
+from Configuration.ProcessModifiers.gpu_cff import gpu
+gpu.toModify(siPixelPhase1MonitorRecHitsSoA, pixelHitsSrc = "siPixelRecHitsPreSplittingSoA") #would be obsloete if .501 is dropped
 
 monitorpixelSoASource = cms.Sequence(siPixelPhase1MonitorRecHitsSoA * siPixelPhase1MonitorTrackSoA * siPixelPhase1MonitorVertexSoA)
+
+## GPU vs CPU validation
+from Configuration.ProcessModifiers.gpuValidationPixel_cff import gpuValidationPixel
```
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/python/SiPixelPhase1HeterogenousDQM_FirstStep_cff.py:16]: @AdrianoDee sorry, but where is this created? 
Btw, it'll conflict with  https://github.com/cms-sw/cmssw/pull/37619 ?

**Comment by AdrianoDee** [comment]
@mmusich you are right, I hadn't pushed the modifications to `Configuration` package. Now it's there. For the possible conflicts we can choose how to proceed and which goes in first. No preference from my side.

@fwyzard may be helpful: find [here](https://adiflori.web.cern.ch/adiflori/37617/dep_graphs/) the dependency graphs for `.502` and `.503` when a GPU is available (or not). The idea is that:
- we have both `siPixelRecHitsPreSplittingCUDA` and `siPixelRecHitsPreSplittingCPU` in the task having only one of the two running (being consumed by the downstream reconstruction);
- the `siPixelRecHitsSoA` is now a `SwitchProducer`(s) that either convert the CUDA to SoA if running on GPU (used only if requested by DQM) either is an alias for `SiPixelRecHitsSoAFromLegacy`

For the suggestion to move `std::is_same` with `std::is_same_v` I've supposed the suggestion was to move it everywhere.

Anyway, I'll squash all the commits at the end of the review.

**Comment by AdrianoDee** [comment]
please test

**Comment by fwyzard** [comment]
> For the possible conflicts we can choose how to proceed and which goes in first. No preference from my side.

I would suggest to merge https://github.com/cms-sw/cmssw/pull/37619 first -- for the only reason that I managed to read through all of it before this one :-)

**Comment by AdrianoDee** [comment]
> > For the possible conflicts we can choose how to proceed and which goes in first. No preference from my side.
> 
> 
> 
> I would suggest to merge https://github.com/cms-sw/cmssw/pull/37619 first -- for the only reason that I managed to read through all of it before this one :-)

Fine by me (: [I'll fix the issue with relvals in the meanwhile]

**Comment by AdrianoDee** [comment]
test parameters:

- enable_tests = gpu
- workflows_gpu = 11634.503

**Comment by AdrianoDee** [comment]
please test

**Comment by AdrianoDee** [comment]
Failures unrelated with this PR and related to https://github.com/cms-sw/cmssw/issues/37584.

**Comment by AdrianoDee** [comment]
test parameters:

- enable_tests = gpu
- workflows_gpu = 11634.503
- pull_request = https://github.com/cms-sw/cmssw/pull/37634

**Comment by AdrianoDee** [comment]
please test

**Comment by emanueleusai** [comment]
@AdrianoDee Can you explain differences in DQM bin by bin GPU comparison in workflow .506?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_4_X_2022-04-20-1100+e624a5/49739/dqm-histo-comparison-summary.html

https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf11634_506_pr/CMSSW_12_4_X-PRcmssw_37617-49739/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf11634_506_base/CMSSW_12_4_X-PRcmssw_37617-49739/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;

**Comment by AdrianoDee** [comment]
@emanueleusai I suspect this comes from the usual irreproducibility of irreproducibility of GPU wfs (similar to what happens in #37619). Looking at the plots the differences are mostly fluctuations.

**Comment by emanueleusai** [comment]
+1

**Comment by emanueleusai** [comment]
+1

File: Configuration/PyReleaseValidation/python/relval_2026.py [hunk 3]
```diff
@@ -34,7 +34,7 @@
 numWFIB.extend([39034.0]) #2026D87
 numWFIB.extend([39434.0,39434.911,39434.103]) #2026D88 DDD XML, DD4hep XML, aging
 numWFIB.extend([39661.97]) #2026D88 premixing stage1 (NuGun+PU)
-numWFIB.extend([39434.5,39434.9,39434.501,39434.502]) #2026D88 pixelTrackingOnly, vector hits, Patatrack local reconstruction on CPU, Patatrack local reconstruction on GPU
+numWFIB.extend([39434.5,39434.9,39434.502]) #2026D88 pixelTrackingOnly, vector hits, Patatrack local reconstruction on CPU, Patatrack local reconstruction on GPU
```
Comment (jpata) [inline Configuration/PyReleaseValidation/python/relval_2026.py:37]: should the comment here be updated as well?

Comment (fwyzard) [reply Configuration/PyReleaseValidation/python/relval_2026.py:37]: I would actually prefer to keep the two configuration (`.501` and `.502`) for the time being

Comment (AdrianoDee) [reply Configuration/PyReleaseValidation/python/relval_2026.py:37]: No problem for me to postpone it. Still would drop it at a certain point (:

**Comment by AdrianoDee** [comment]
test parameters:

- enable_tests = gpu
- workflows_gpu = 11634.503, 11634.502

**Comment by AdrianoDee** [comment]
please test

**Comment by AdrianoDee** [comment]
Latest commit(s):
- fixing conflicts and realigning to https://github.com/cms-sw/cmssw/pull/37619 ;
- restoring `.501`;
- updating description and title accordingly.

**Comment by jpata** [comment]
+reconstruction
- no reco changes (DQM-related reorganization)

**Comment by jpata** [comment]
+reconstruction
- no reco changes (DQM-related reorganization)

**Comment by fwyzard** [comment]
Thanks @AdrianoDee for the dependency graphs.

Here is a comparison of the `.502` workflow running only on CPU before this PR
502_cpu
and after:
502_cpu

The difference is that the `siPixelRecHitsPreSplittingSoA` module has been replaced by a `SwitchProducer`, and the module doing the reconstruction on CPU now is `siPixelRecHitsPreSplittingCPU`.

Here is the same comparison for the `.502` workflow running on GPU, before

and after


The `siPixelRecHitsPreSplittingSoA` module hre is replaced by the `@cuda` branch of the `SwitchProducer`, which runs the `SiPixelRecHitSoAFromCUDA` module instead of the `SiPixelRecHitSoAFromLegacy` one.

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
@AdrianoDee with these changes, how does the `.502` workflow running on CPU compare with the `.501` workflow ?

**Comment by perrotta** [comment]
+1

**Comment by AdrianoDee** [comment]
@fwyzard, I checked the dependency graphs ([in pdf here](https://github.com/cms-sw/cmssw/files/8611475/501vs502.pdf)) and the tasks and there is a misalignment in the names in the local reco:

- in `.501` the `siPixelRecHitsPresplitting` holds the legacy and the SoA and it is a `SwitchProducedCUDA` with the `@cpu` branch only;

- in `.502` on GPU the `siPixelRecHitsPresplitting` holds only the legacy hits (either through and `EDAlias` for `@cpu` or with the `SiPixelRecHitFromCUDA` new producer) and the SoA is in `siPixelRecHitsPreSplittingSoA` (again either through and `EDAlias` for `@cpu` or with the `SiPixelRecHitsSoAFromCUDA` new producer), being both a `SwitchProducedCUDA` with both branches;

For the rest (and the final result) they run the same. Is this naming mismatch a possible issue? In case I'm trying to figure out a possible workaround given that, as far as I have understood `SwitchProducedCUDA` can't hold a single branch with an `EDAlias` and a producer may not be _replaced with_ with an `EDAlias`.
 
501vs502

**Comment by AdrianoDee** [comment]
Also, do we need a backport of this? I missed the ORP today, sorry.

**Comment by fwyzard** [comment]
> Also, do we need a backport of this? I missed the ORP today, sorry.

No, I don't think so.

**Comment by fwyzard** [comment]
> `SwitchProducedCUDA` can't hold a single branch with an `EDAlias`

Actually, I don't that would be a problem.

**Comment by qliphy** [comment]
@AdrianoDee   There is a IB issue after this PR gets merged, would you please have a check?
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc10/CMSSW_12_4_X_2022-05-03-2300/pyRelValMatrixLogs/run/39434.501_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal.log#/

cmsRun: /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/5525f5050df49f51ff7c29cd32fd5dea/opt/cmssw/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_X_2022-05-03-2300/src/RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:144: virtual void SiPixelRecHitSoAFromLegacy::produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const: Assertion `gind < nMaxModules' failed.
cmsRun: /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/5525f5050df49f51ff7c29cd32fd5dea/opt/cmssw/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_X_2022-05-03-2300/src/RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:144: virtual void SiPixelRecHitSoAFromLegacy::produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const: Assertion `gind < nMaxModules' failed.

**Comment by AdrianoDee** [comment]
> @AdrianoDee   There is a IB issue after this PR gets merged, would you please have a check?
> 
> https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc10/CMSSW_12_4_X_2022-05-03-2300/pyRelValMatrixLogs/run/39434.501_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal.log#/
> 
> 
> 
> cmsRun: /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/5525f5050df49f51ff7c29cd32fd5dea/opt/cmssw/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_X_2022-05-03-2300/src/RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:144: virtual void SiPixelRecHitSoAFromLegacy::produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const: Assertion `gind < nMaxModules' failed.
> 
> cmsRun: /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/5525f5050df49f51ff7c29cd32fd5dea/opt/cmssw/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_X_2022-05-03-2300/src/RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:144: virtual void SiPixelRecHitSoAFromLegacy::produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const: Assertion `gind < nMaxModules' failed.

@qliphy on it

**Comment by AdrianoDee** [comment]
@qliphy fixed in #37798

> Actually, I don't know if that would be a problem -- is it something you tried and didn't work ?

@fwyzard I was doing some other stupid mistake. It actually works and I implemented in #37798

**Pull request merged**